submissions/nadsatt/post-oop/index.js

```diff
@@ -0,0 +1,152 @@
+            but i'm not sure about both ideas because it makes my code strange a bit (especially when i need to create FriendsManager instance for each inhabitant)
+   - FOR O: i changed 'toString' method in 'Inhabitant' to make it dependent on specific properties 
+            to have ability to redefine this method in subclasses
+            to show how this method can be simultaneously closed for modifications in superclass and open for modifications in subclasses
+   - FOR L: as i see, there is no cases where subclass can't be used instead of its superclass
+            for example, in 'addFriends' method of 'FriendsManager' class instance of any subclass can be used
+   - FOR I: as i see, there is no unused/unrelevant superclass' methods/props in subclasses
+   - FOR D: -
+*/ 
+
+class Inhabitant {
+   constructor(species, saying, gender, name, friends = []){
+      this.species = species;
+      this.saying = saying;
+      this.gender = gender;
+      this.name = name;
+      this.friends = friends;
+   }
+
+   toString(){
+      return `species: ${this.species}, saying: ${this.saying}, gender: ${this.gender}, name: ${this.name}, friends: ${new FriendsManager(this).getFriends()}`;
+   }
+}
+
+class FriendsManager {
+   constructor(inhabitant){
+      this.inhabitant = inhabitant;
+   }
+
+   addFriends(...friends){
+      this.inhabitant.friends.push(...friends);
+   }
+
+   getFriends(){
+      return this.inhabitant.friends.map(friend => friend.name).join(' ') || 'no friends';
+   }
+}
+
+class Primate extends Inhabitant {
+   constructor(species, saying, gender, name, ...friends){
+      super(species, saying, gender, name, friends);
+      this.hands = 2;
+      this.legs = 2;
+   }
+
+   toString(){
+      return super.toString() + `, legs: ${this.legs}, hands: ${this.hands}`;
+   }
+}
+
+class Human extends Primate {
+   constructor(saying, gender, name, ...friends){
+      super('human', saying, gender, name, ...friends);
+   }
+}
+
+class Man extends Human {
+   constructor(name, ...friends){
+      super('i\'m a man!', 'male', name, ...friends);
+   }
+}
+
+class Woman extends Human {
+   constructor(name, ...friends){
+      super('i\'m a woman!', 'female', name, ...friends);
+   }
+}
+
+class Predator extends Inhabitant {
+   constructor(species, saying, gender, name, ...friends){
+      super(species, saying, gender, name, friends);
+      this.legs = 4;
+   }
+
+   toString(){
+      return super.toString() + `, legs: ${this.legs}`;
+   }
+}
+
+class Dog extends Predator {
+   constructor(gender, name, ...friends){
+      super('dog', 'woof-woof!', gender, name, ...friends);
+   }
+}
+
+class Cat extends Predator {
+   constructor(gender, name, ...friends){
+      super('cat', 'meov-meov!', gender, name, ...friends);
+   }
+}
+
+class CatWoman extends Cat {
+   constructor(name, ...friends){
+      super('female', name, ...friends);
+      this.species = 'human';
+      this.hands = 2;
+      this.legs = 2;
```

Comment:
**DO NOT FIX** anything as per the comment below.

Well, creation of hybrids in traditional class-syntax approach is always a challenge.

Compositional pattern would work much better but that would require you to rewrite the entire codebase into functional style.

You do need to fix anything here. Just FYI to be careful and whenever you see space for hybrids consider switching to compositional pattern instead of classic inheritance pattern.